reading-writing-files.md

```diff
@@ -19,41 +19,91 @@ access commonly used locations on the device's filesystem. The class currently s
 
-## Writing a string to a file
-Use the File object’s [writeAsString()](https://api.dartlang.org/stable/1.18.1/dart-io/File/writeAsString.html) method to write a string to a file. After writing the string, the method closes the file.
+## Example of reading and writing to a file
 
-```dart
-import 'dart:io';
-import 'dart:async';
-import 'package:flutter/services.dart';
-import 'package:flutter/widgets.dart';
-
-Future<Null> main() async {
-  runApp(null); // this line is required to initialize the shell
-  String dir = (await PathProvider.getApplicationDocumentsDirectory()).path;
-  String filename = "$dir/foo.txt";
-  File file = new File(filename);
-  await file.writeAsString('some content');
-}
-```
-
-## Reading a file as a string
-Use the File object's [`readAsString()`](https://api.dartlang.org/stable/1.18.1/dart-core/Resource/readAsString.html) method to read a file as a string, then display it on the screen.
+The following example shows how to modify the default app created by `flutter create`, so that the number of button presses are peristsed across relaunches of the app.
 
 ```dart
 import 'dart:io';
 import 'dart:async';
 import 'package:flutter/services.dart';
-import 'package:flutter/widgets.dart';
-
-Future<Null> main() async {
-  runApp(null); // this line is required to initialize the shell
-  String dir = (await PathProvider.getApplicationDocumentsDirectory()).path;
-  String filename = "$dir/foo.txt";
-  File file = new File(filename);
-  String contents = await file.readAsString();
+import 'package:flutter/material.dart';
+
+void main() {
   runApp(
-    new Center(child: new Text('Contents: $contents'))
+    new MaterialApp(
+      title: 'Flutter Demo',
+      theme: new ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      home: new FlutterDemo(),
+    ),
   );
 }
+
+class FlutterDemo extends StatefulWidget {
+  FlutterDemo({Key key}) : super(key: key);
+
+  @override
+  _FlutterDemoState createState() => new _FlutterDemoState();
+}
+
+class _FlutterDemoState extends State<FlutterDemo> {
+  int _counter;
+
+  @override
+  void initState() {
+    super.initState();
+    _readCounter();
+  }
+
+  Future<File> _getLocalFile() async {
+    String dir =
+        (await PathProvider.getApplicationDocumentsDirectory()).path;
+    String filename = "$dir/counter.txt";
+    return new File(filename);
+  }
+
+  Future<Null> _readCounter() async {
+    try {
+      File file = await _getLocalFile();
+      String contents = await file.readAsString();
+      setState(() {
+        _counter = int.parse(contents);
+      });
+    } on FileSystemException {
+      setState(() {
+        _counter = 0;
+      });
+    }
+  }
+
+  void _incrementCounter() {
+    setState(() {
+      _counter++;
+      _getLocalFile().then((File file) {
+        file.writeAsStringSync("$_counter");
```

Comment:
Talking to @Hixie we would like to make this an asynchronous call, and move it outside of the setState() function. We'd like to train developers that they should do async I/O.